File: Problem_3.js
// ## Problem3  Rotate Array by K Places(https://leetcode.com/problems/rotate-array/)

// Time Complexity : O(n logn)
// Space Complexity : O(1)

// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

var reverse = (nums, l, r) => {
    while (l < r) {
        let temp = nums[l];
        nums[l] = nums[r];
        nums[r] = temp;
        l++;
        r--;
    }
}
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function (nums, k) {
    let n = nums.length;
    if (k >= n) {
        k = k % n;
    }
    reverse(nums, 0, n - 1)
    reverse(nums, 0, k - 1)
    reverse(nums, k, n - 1)
    return nums;
};

File: Problem3.cs
public class RotateArrayKTimes
	{
        // Time Complexity : O(n)
        // Space Complexity : O(1) - in place
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        public void Rotate1(int[] nums, int k)
        {
            if (nums == null || nums.Length == 0) return;
            int n = nums.Length;
            if(k > n)
            {
                k = k % n;
            }

            Reverse(nums, 0, n - 1);
            Reverse(nums, 0, k - 1);
            Reverse(nums, k, n - 1);
 
        }

        public void Reverse(int[] nums, int l, int r)
        {
            while(l < r)
            {
                Swap(nums, l, r);
                l++;
                r--;
            }
        }

        public void Swap(int[] arr, int i, int j)
        {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
}
